### DISCUSSION TITLE
Can I share functions across modules?

### DISCUSSION THREAD

**Discussion by wsw70** [custom-components/pyscript#286] [General] [answered]
I used to have function split "by functionality" (a module for lights, another one for the roller shutters, ...) and I am now reorganizing them "by room". More code but also more readable.

I have therefore several functions such as the one below where I am iterating through several "rooms" (in the example below - just one - `parents`).

```python
@time_trigger("once(09:00)")
def volets_matin_partiel():
    partial_height = 40
    for who in ['parents']:
        if state.get(f"input_boolean.volets_{who}") == 'on':
            if state.getattr(f"cover.{who}_volets")['current_position'] < partial_height:
                cover.set_cover_position(entity_id=f"cover.{who}_volets", position=partial_height)
            else:
                log.info(f"volets {who} déjà ouverts")
```

I could duplicate them across the "rooms" but that would be repeated code and I would prefer to keep one function (say, in a module `libraries.py` and call them with the relevant parameters (the name of the room for instance).

How should I approach this?

In other words, I would like something in room "michael" (in file `michael.py`)  such as

```python
@time_trigger("once(09:00)")
volets_partiel(room=michael, partiel=40)
```

This is not the correct code because the decorator is for function definition (and not function call), I am not sure if there is a way?

**Comment by dlashua** [comment, marked as answer]
I use the `modules` directory for this, and pyscript has some stuff builtin to help.

modules/volets_helpers.py
```
def volets_partiel(blah, blah):
  @time_trigger("once(09:00)")
  def inner(room=None, partiel=0):
    blah blah

  return inner
```

scripts/michael.py
```
from volets_helpers import volets_partiel

triggers = []
triggers.append(volets_partiel(room=michael, partiel=40))
```

**Reply by wsw70** [reply]
Thanks a lot. I am trying to understand your code:
- how come you pass `room=michael` (and not `blah`) to `volets_partiel` and it lands in `inner`?
- what is the `triggers` list?

**Reply by dlashua** [reply]
Sorry... I didn't pay much attention to the parameters of the functions I was only trying to show the pattern.

which parameters go where will depend entirely upon the code in `volets_partiel`.

The triggers list is a way of making sure the closure is registered with pyscript. [more info here](https://hacs-pyscript.readthedocs.io/en/stable/reference.html#trigger-closures)

**Reply by wsw70** [reply]
Below is a working example, just to show the use of outer and inner parameters (something I did not catch initially)

```
def volets_time(when='', entity_id=None, what=None, condition='on'):
    # all the parameters can now be used below, including in decorators and inner function
    log.info(f"setting up volets for {entity_id} at {when} → {what}")
    @time_trigger(f"once({when})")  # setting up a "dynamic" time trigger
    def volets():
        if condition == 'on':
            if what == "up":
                operation = cover.open_cover(entity_id=entity_id)
            else:
                operation = cover.close_cover(entity_id=entity_id)
            log.info(f"volets {entity_id} → {what}")
        else:
            log.info(f"volets {entity_id} should go {what}, but automation is switched off")
    return volets
```
